ochre/Equipment/PV.py

```diff
@@ -1,71 +1,148 @@
+            roofs = [bd.ext_surface for bd in envelope_model.boundaries if 'Roof' in bd.name]
+            # Use roof closest to south with preference to west (0-45 degrees)
+            roof_data = pd.DataFrame([[bd.tilt, az] for bd in roofs for az in bd.azimuths], columns=['Tilt', 'Az'])
+            best_idx = (roof_data['Az'] - 185).abs().idxmax()
+            self.tilt = roof_data.loc[best_idx, 'Tilt']
+            self.azimuth = roof_data.loc[best_idx, 'Az']
+
+        # Inverter constraints
+        self.inverter_capacity = inverter_capacity or self.capacity  # in kVA, AC
+        self.inverter_efficiency = inverter_efficiency
+        self.inverter_priority = inverter_priority
+        self.inverter_min_pf = inverter_min_pf
+        if self.inverter_min_pf is not None:
+            self.inverter_min_pf_factor = ((1 / self.inverter_min_pf ** 2) - 1) ** 0.5
+        else:
+            self.inverter_min_pf_factor = None
 
-            # Note: defining schedule as positive=consuming power
-            schedule = schedule['Power_AC'] if include_inverter else schedule['Power_DC']
-            schedule *= capacity  # normalize with capacity, in kW
-            schedule = schedule.to_frame(self.name + ' (kW)')
-            kwargs['schedule'] = schedule
+        super().__init__(envelope_model=envelope_model, **kwargs)
 
-        super().__init__(**kwargs)
+        self.q_set_point = 0  # in kW, positive = consuming power
+        if self.capacity is None:
+            self.capacity = -self.schedule[self.electric_name].min()
+        if self.inverter_capacity is None:
+            self.inverter_capacity = self.capacity
 
         # check that schedule is negative
-        if abs(self.schedule[self.electric_name].min()) < abs(self.schedule[self.electric_name].max()):
+        if self.schedule[self.electric_name].abs().min() < self.schedule[self.electric_name].abs().max():
             self.warn('Schedule should be negative (i.e. generating power).',
                       'Reversing schedule so that PV power is negative/generating')
-            self.schedule = -self.schedule
+            self.schedule = self.schedule * -1
             self.reset_time()
 
-        self.capacity = capacity if capacity is not None else -self.schedule[self.electric_name].min()  # in kW, DC
-        self.q_set_point = 0  # in kW, positive = consuming power
+    def initialize_schedule(self, schedule=None, equipment_schedule_file=None, location=None, **kwargs):
+        if (schedule is None or self.name + ' (kW)' not in schedule) and equipment_schedule_file is None:
+            self.print('Running SAM')
+            schedule = run_sam(self.capacity, self.tilt, self.azimuth, schedule, location,
+                               self.inverter_capacity, self.inverter_efficiency)
+            schedule = schedule.to_frame(self.name + ' (kW)')
 
-        # Inverter constraints, if included; if not, assume no limits and 100% efficiency
```

Comment:
Rather than 100% efficiency, should we try to pick something that lines up with the SAM defaults for inverter efficiency? Or is inverter efficiency included in the PV inefficiencies so that would be double counting?

Reply:
This note is old, we are using SAM's inverter efficiency value, unless the
user specifies something else